### DISCUSSION TITLE
How does nushell know some output is a table?

### DISCUSSION THREAD

**Discussion by ghost** [nushell/nushell#12036] [General]
Apologies if this is extremely basic - I skimmed parts of the book, saw some mention of 'external commands' and so on but didn't answer my question along the way...

I'm trying nushell for the first time after seeing a post about it on lobste.rs, and I'm wondering how nushell "knows" when to print the very nice looking table outputs (or use other table-related features like sorting).

For example:
- `ls` and `ps` print tables (as shown in the docs/book)
- `exa -l` does not print with a table (even though the information is, to me as a naive user, roughly the same table of values as `ls -l`)
- `docker images` does not print with a table (again, it's obviously table-style structured output given by the command)
- if I make a copy of `ls` and call it `asdf`, running `asdf` does not have the same table output as running `ls`

Is there something like a lookup table implementing this behind the scenes?

Is there a future possibility of "detecting table-style output" to make it more general?

**Comment by fdncred** [comment]
nushell has its own structured data format that it supports, the internal `ls` and `ps` commands understand this format. external tools like `exa` and `docker` have no idea about that structured data format. we provide tools to parse external data into nushell's tabular output format. Those are mostly `parse`, `from ssv`, and `detect columns`. You can also use any tool that outputs json and pipe that into `from json` or `from json -o` if it is jsonl/ndjson formatted.

**Reply by ghost** [reply]
Gotcha, that makes sense. I started discovering this by running `which` (seeing what it labels as built-in, including itself). 

Cool - thanks for clarifying!

**Reply by develop7** [reply]
@fdncred Is there a way for an external command to output the data in nushell's native format?

**Reply by ghost** [reply]
random executables can't create nushell data afaik, but there are a few options (all of which require configuration on the shell side):
* **plugins** have to be registered, but can be written in any language and take "native" (or as close as well get) nushell data as in and output
* the **`extern` command** can be used to explain to nushell how to use an external command and simple types like `int` and `path` can be marked to help with completion, etc (and its also possible to write complex tab-completions)
* you can write nushell **wrapper functions** (`def`), which shadow (aka override) external (sub-)commands and parse the (for example json) output. (nuon would be a more native format to parse than json especially since it supports nushells time format, etc, but in my experience its usually easier to just use json since there are libraries everywhere. (also there are some instances where the nuon parser has some problems for example issue #9879 ))

**Reply by fdncred** [reply]
You could also build nushell into your command as you build it in Rust, by using our crates. I've seen a few of these like [nur](https://github.com/ddanier/nur), [quake](https://github.com/quake-build/quake), and [couchbase-shell](https://github.com/couchbaselabs/couchbase-shell).
